File: cli/internal/api/pow.go
package api

import (
	"crypto/sha256"
	"encoding/hex"
	"math/bits"
	"strconv"
)

// RegisterChallenge is the response from GET /v1/auth/register-challenge.
type RegisterChallenge struct {
	Nonce      string `json:"nonce"`
	Difficulty uint   `json:"difficulty"`
	ValidUntil string `json:"valid_until"`
}

// leadingZeroBits counts leading zero bits in the hash (from the high byte).
func leadingZeroBits(hash [32]byte) uint {
	var n uint
	for _, b := range hash {
		if b == 0 {
			n += 8
		} else {
			n += uint(bits.LeadingZeros8(b))
			break
		}
	}
	return n
}

// SolvePoW finds a solution such that SHA256(nonce_bytes || valid_until_utf8 || solution_utf8)
// has at least difficulty leading zero bits. Returns the solution string.
func SolvePoW(nonceHex, validUntil string, difficulty uint) (string, error) {
	nonce, err := hex.DecodeString(nonceHex)
	if err != nil {
		return "", err
	}
	validUntilBytes := []byte(validUntil)
	for trial := uint64(0); ; trial++ {
		solution := strconv.FormatUint(trial, 10)
		h := sha256.New()
		h.Write(nonce)
		h.Write(validUntilBytes)
		h.Write([]byte(solution))
		var hash [32]byte
		copy(hash[:], h.Sum(nil))
		if leadingZeroBits(hash) >= difficulty {
			return solution, nil
		}
	}
}

File: e2e/common/pow.py
import hashlib


def leading_zero_bits(hash_bytes: bytes) -> int:
    """
    Count leading zero bits in the hash, from MSB of the first byte.
    Matches the semantics of the backend PoW difficulty checks.
    """

    n = 0
    for b in hash_bytes:
        if b == 0:
            n += 8
        else:
            # number of leading zeros within the byte
            n += 8 - b.bit_length()
            break
    return n


def solve_pow(nonce_hex: str, valid_until: str, difficulty: int) -> str:
    """
    Finds a solution string such that:
      SHA256(nonce_bytes || valid_until_utf8 || solution_utf8)
    has at least `difficulty` leading zero bits.
    """

    nonce = bytes.fromhex(nonce_hex)
    vu_bytes = valid_until.encode("utf-8")
    prefix = nonce + vu_bytes

    trial = 0
    difficulty = int(difficulty)
    while True:
        solution = str(trial).encode("utf-8")
        h = hashlib.sha256(prefix + solution).digest()
        if leading_zero_bits(h) >= difficulty:
            return solution.decode("utf-8")
        trial += 1

